index.rst

```diff
@@ -1,61 +1,173 @@
+
+Data Sizing Calculations
+------------------------
+
+DMTN-102 :cite:`DMTN-102` provides estimates for the total volume of alerts received.
+We expect a maximum of about 10 million alerts per night, each about 80 kilobytes.
+LSE-163 :cite:`LSE-163` projects about 300 nights of observations per year.
+
+Combining these figures, we project 80 gigabytes of alert data per night, and 2.4 terabytes per year.
+
+Existing Requirements
+---------------------
+
+The database is referenced in a handful of existing project requirements documents:
+
+ - OSS-REQ-0128 "Alerts" :cite:`LSE-30`:
+
+     The Level 1 Data Products shall include the Alerts produced as part  of the nightly Alert Production.
+
+ - OSS-REQ-0185 "Transient Alert Query" :cite:`LSE-30`:
+
+     All published transient alerts, as well as all reprocessed historical alerts generated as part of a Data Release, shall be available for query.
+
+ - DMS-REQ-0094 "Keep Historical Alert Archive" :cite:`LSE-61`:
 
-.. TODO: Delete the note below before merging new content to the master branch.
+     The DMS shall preserve and keep in an accessible state an alert archive with all issued alerts for a historical record and for false alert analysis.
+
+ - DMS-PRTL-REQ-0033 "Queries on the Alerts Database" :cite:`LDM-554`:
+
+     The Portal aspect shall provide a query interface to the Alert Database, allowing searches based on parameters which shall include, but may not be limited to, Alert ID, time of alert, position on the sky, filter, and alert characteristics.
+
+ - DMS-PRTL-REQ-0048 "Alert Visualization" :cite:`LDM-554`:
+
+     The Portal aspect shall provide for the users a "property sheet" for the contents of an alert packet including, but not necessarily limited to, the alert postage stamp image, the postage stamp time series, the photometric time series, the source and object information (e.g., position, brightness).
+
+Proposed Implementation
+=======================
+
+We can satisfy these design inputs by storing serialized Avro alert data (the same bytes sent via Kafka to brokers) in a S3-like object store, indexed by a unique alert ID.
+Each alert packet corresponds to one object in the object store.
 
 .. note::
 
-   **This technote is not yet published.**
+   An alternative would be to combine many packets into a block in the object store, perhaps of about 100 alert packets.
+   This might permit more efficient storage and retrieval.
+   Storage might be more efficient because compression would be better when storing many alerts.
+   Retrieval might be more efficient because it might tune the outgoing flows into a smaller number of TCP connections which get a chance to grow window sizes.
 
-   Design document for a database of all alerts sent by the Rubin observatory.
+   But this would be more complex, and make writing more difficult, as writes need to append to existing data which would require coordination between writers, so this design sticks to a simpler structure.
```

Comment:
Another question raised by the discussion on DM-28734 is whether the alerts are stored on disk with schemas.  We are not transmitting alerts with schemas for bandwidth reasons, so a verbatim archive of what was transmitted would not include the schemas.  However, if we archive the alerts with the schemas users (and our own services) will not need to retrieve the appropriate schemas from elsewhere.

Since Avro schema storage is quite inefficient, archiving alerts with schemas might push us somewhat in the direction of a single file (with schema) per CCD.  

If we did store multiple alerts as a single object, in order to retrieve a single alert the alertDB service would have to seek through the file to find it, which would add latency.   That could hinder using the alertDB as a key component of a hybrid alert system.

Reply:
Interesting point, yeah.

The space savings probably approaches 1.5x to 2x, which doesn't seem like enough to really be a decider.

A more compelling reason to do this, though, would be that it reduces CPU burden on callers. Parsing the entire schema document for _every_ file is very wasteful; parsing schemas is usually not very well-optimized in Avro libraries since it's expected that it's done just once. We could remediate that with a bit of extra complexity: we could run a fronting service which strips schemas from alerts when they're requested.

As you say, an alertDB service would need to seek for a specific alert in a file to respond with it. But how much latency would this add? 

A single CCD-visit is expected to have about `10,000 / 189 = 52` alerts. How long does it take to scan 52 alerts?

I don't know exactly, but I can estimate based on [Avro benchmarks of ZTF data](https://github.com/fastavro/fastavro/issues/503) run on my laptop. A naive decoder with the `fastavro` library takes about 1.5ms per alert; a tuned hand-written one takes about 01ms per alert. So we're looking at adding something like 50ms to 70ms - which is actually quite significant, much more than I would have ever expected.

Reply:
Keep in mind that it is possible for the alert schema to evolve over time, which might affect how schemas are stored and transmitted.

Reply:
Right, schema evolution is a very good reason to store schemas directly in files. It's actually [mandated in the Avro specification for that reason](http://avro.apache.org/docs/current/spec.html#Data+Serialization+and+Deserialization):

>  files or systems that store Avro data should always include the writer's schema for that data.

To my mind, there's no real question that we should include the schema in the files we store. But we could amortize the overhead of that inclusion by storing many alerts in a single file, like Eric says. I think we'd need to record byte offsets of each sub-alert in order to make that efficient though.

---

One unorthodox way we could do that would be to include a mapping of `alertID -> integer index of the alert within the file`. This could be included as a `map`-typed value in the metadata header of an Avro Object Container file.

Next, each alert could be stored a single "data block" within the Avro Object Container file. Data blocks are always prefixed with their size in bytes, which lets readers efficiently skip entire blocks.

Then, the seeking will be pretty close to free. We read (and parse) the Avro schema which is embeded in the file, and then read the header metadata to figure out how many blocks to skip. We skip each block based on the byte size field, and then actually extract the alert packet only when we have hit the block we're seeking for.

This is fairly complex stuff and uses parts of Avro that are rarely exposed in Python libraries, but I don't think it's very difficult really, and the complexity could be well-encapsulated, either in a Python module that people use (an AlertDB API client, I guess), or a remote service that does this on their behalf.

Reply:
What do you think about the above? The tradeoff is essentially:

 - More code complexity
 - Less disk usage in the object store (The 50KB schema is stored once per ~50 images instead of once per image, so saves about 50KB per alert, which is about 20TB per year.

Bandwidth savings are a wash: we would save sending the schema, but need to send a bunch of superfluous alerts when plucking a single one out, even if we skip them.

CPU savings are probably a wash too.

Reply:
If we're using a custom reader that is not simply returning the object in the object store (so that there's no need for that object to conform to Avro specifications), then I think it's far simpler to store the schema in a detached form (once per schema evolution change) and attach it to the object upon reading.

But it would be nice to not need the custom reader (which then requires the schema to be attached directly).  Across-object compression isn't a thing, I guess...

Reply:
I've revised to say that we should store schemas separately, using a 4-byte ID.